File: recamera-intellisense-sdk/recamera_intellisense/storage.py
"""Storage slots, quota, and async tasks (``/record/storage/{status,config,control}``)."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from . import _config, _http

__all__ = [
    "get_storage_status",
    "set_storage_slot",
    "configure_storage_quota",
    "storage_task_submit",
    "storage_task_status",
    "storage_task_cancel",
    "ensure_storage",
    "normalize_action",
    "DEFAULT_INTERNAL_DEV_PATH",
]

PATH_STATUS = "/cgi-bin/entry.cgi/record/storage/status"
PATH_CONFIG = "/cgi-bin/entry.cgi/record/storage/config"
PATH_CONTROL = "/cgi-bin/entry.cgi/record/storage/control"

DEFAULT_INTERNAL_DEV_PATH = "/dev/mmcblk0p8"

_STORAGE_ACTIONS = {
    "FORMAT",
    "FREE_UP",
    "EJECT",
    "REMOVE_FILES_OR_DIRECTORIES",
}
_STORAGE_ACTION_ALIASES = {
    "REMOVE": "REMOVE_FILES_OR_DIRECTORIES",
    "REMOVE_FILES": "REMOVE_FILES_OR_DIRECTORIES",
}


def normalize_action(action: str) -> str:
    """Canonicalise a storage-task action name (accepts aliases)."""
    up = str(action).upper()
    if up in _STORAGE_ACTIONS:
        return up
    if up in _STORAGE_ACTION_ALIASES:
        return _STORAGE_ACTION_ALIASES[up]
    raise ValueError(
        f"Unknown storage action {action!r}. Allowed: {sorted(_STORAGE_ACTIONS)}."
    )


def _parse_slot(s: Dict[str, Any], data_dir: str) -> Dict[str, Any]:
    def _s(k: str) -> str:
        return str(s.get(k, ""))

    def _b(k: str) -> bool:
        return bool(s.get(k, False))

    def _i(k: str) -> int:
        return int(s.get(k, 0) or 0)

    return {
        "dev_path": _s("sDevPath"),
        "mount_path": _s("sMountPath"),
        "removable": _b("bRemovable"),
        "internal": _b("bInternal"),
        "label": s.get("sLabel"),
        "uuid": s.get("sUUID"),
        "fs_type": s.get("sType"),
        "selected": _b("bSelected"),
        "enabled": _b("bEnabled"),
        "syncing": _b("bSyncing"),
        "writing": _b("bWriting"),
        "rotating": _b("bRotating"),
        "state_code": _i("eState"),
        "state": _s("sState"),
        "size_bytes": _i("iStatsSizeBytes"),
        "free_bytes": _i("iStatsFreeBytes"),
        "quota_min_recommend_bytes": _i("iQuotaMinimumRecommendBytes"),
        "quota_preserved_bytes": _i("iQuotaPreservedBytes"),
        "quota_used_bytes": s.get("iQuotaUsedBytes"),
        "quota_limit_bytes": _i("iQuotaLimitBytes"),
        "quota_rotate": _b("bQuotaRotate"),
        "data_dir": data_dir,
    }


def get_storage_status(device_name: str) -> List[Dict[str, Any]]:
    """List the storage slots (each as a dict)."""
    dev = _config.resolve(device_name)
    data = _http.get_json(dev, PATH_STATUS) or {}
    data_dir = data.get("sDataDirName", "") or ""
    slots = data.get("lSlots") or []
    return [_parse_slot(s, data_dir) for s in slots if isinstance(s, dict)]


def set_storage_slot(
    device_name: str,
    *,
    by_dev_path: str = "",
    by_uuid: str = "",
) -> None:
    """Select the slot to enable; pass both empty to disable all slots."""
    dev = _config.resolve(device_name)
    if by_dev_path or by_uuid:
        select: Any = {"sByDevPath": by_dev_path, "sByUUID": by_uuid}
    else:
        select = None
    resp = _http.post_json(dev, PATH_CONFIG, payload={"dSelectSlotToEnable": select})
    _http.expect_ok(resp, "set storage selection")


def configure_storage_quota(
    device_name: str,
    *,
    dev_path: str,
    quota_limit_bytes: int,
    quota_rotate: bool = True,
) -> None:
    """Set per-slot quota. ``quota_limit_bytes = -1`` means no limit."""
    dev = _config.resolve(device_name)
    payload = {
        "sTaskType": "SYNC",
        "sAction": "CONFIG",
        "sSlotDevPath": dev_path,
        "dSlotConfig": {
            "iQuotaLimitBytes": int(quota_limit_bytes),
            "bQuotaRotate": bool(quota_rotate),
        },
    }
    resp = _http.post_json(dev, PATH_CONTROL, payload=payload)
    _http.expect_ok(resp, "configure storage quota")


def _task_payload(
    task_type: str,
    action: str,
    dev_path: str,
    files: Optional[List[str]] = None,
) -> Dict[str, Any]:
    action = normalize_action(action)
    payload: Dict[str, Any] = {
        "sTaskType": task_type,
        "sAction": action,
        "sSlotDevPath": dev_path,
    }
    if action == "REMOVE_FILES_OR_DIRECTORIES":
        if not files:
            raise ValueError("REMOVE_FILES_OR_DIRECTORIES requires non-empty 'files'.")
        payload["lFilesOrDirectoriesToRemove"] = list(files)
    return payload


def storage_task_submit(
    device_name: str,
    *,
    action: str,
    dev_path: str,
    sync: bool = False,
    files: Optional[List[str]] = None,
) -> Dict[str, Any]:
    """Submit a storage action. ``sync=True`` blocks until completion."""
    dev = _config.resolve(device_name)
    payload = _task_payload("SYNC" if sync else "ASYNC_SUBMIT", action, dev_path, files)
    resp = _http.post_json(dev, PATH_CONTROL, payload=payload)
    _http.expect_ok(resp, f"storage task submit {payload['sAction']}")
    return resp if isinstance(resp, dict) else {}


def storage_task_status(
    device_name: str,
    *,
    action: str,
    dev_path: str,
) -> Dict[str, Any]:
    """Query the status of the last async task for (*action*, *dev_path*)."""
    dev = _config.resolve(device_name)
    payload = _task_payload("ASYNC_STATUS", action, dev_path)
    resp = _http.post_json(dev, PATH_CONTROL, payload=payload)
    _http.expect_ok(resp, f"storage task status {payload['sAction']}")
    return resp if isinstance(resp, dict) else {}


def storage_task_cancel(
    device_name: str,
    *,
    action: str,
    dev_path: str,
) -> Dict[str, Any]:
    """Cancel an in-flight async task."""
    dev = _config.resolve(device_name)
    payload = _task_payload("ASYNC_CANCEL", action, dev_path)
    resp = _http.post_json(dev, PATH_CONTROL, payload=payload)
    _http.expect_ok(resp, f"storage task cancel {payload['sAction']}")
    return resp if isinstance(resp, dict) else {}


def ensure_storage(device_name: str, *, timeout_s: float = 3.0) -> None:
    """Ensure one slot is enabled with rotate-quota on (mirrors Rust ``storage::ensure_storage``)."""
    import time

    slots = get_storage_status(device_name)
    if not any(s["enabled"] for s in slots):
        default = next(
            (s for s in slots if s["dev_path"] == DEFAULT_INTERNAL_DEV_PATH), None
        )
        if default is None:
            raise ValueError(
                f"Default storage '{DEFAULT_INTERNAL_DEV_PATH}' not found; "
                "call set_storage_slot to pick one."
            )
        set_storage_slot(device_name, by_dev_path=default["dev_path"])
        deadline = time.time() + timeout_s
        while time.time() < deadline:
            current = get_storage_status(device_name)
            match = next(
                (s for s in current if s["dev_path"] == default["dev_path"]), None
            )
            if (
                match
                and match["enabled"]
                and match.get("state") in ("IDLE", "WRITING", "READY")
            ):
                break
            time.sleep(0.25)
        configure_storage_quota(
            device_name,
            dev_path=default["dev_path"],
            quota_limit_bytes=-1,
            quota_rotate=True,
        )
        return
    for s in slots:
        if s["enabled"] and not s["quota_rotate"]:
            configure_storage_quota(
                device_name,
                dev_path=s["dev_path"],
                quota_limit_bytes=s["quota_limit_bytes"],
                quota_rotate=True,
            )


COMMANDS = {
    "get_storage_status": get_storage_status,
    "set_storage_slot": set_storage_slot,
    "configure_storage_quota": configure_storage_quota,
    "storage_task_submit": storage_task_submit,
    "storage_task_status": storage_task_status,
    "storage_task_cancel": storage_task_cancel,
}
COMMAND_SCHEMAS = {
    "get_storage_status": {"required": {"device_name"}, "optional": set()},
    "set_storage_slot": {
        "required": {"device_name"},
        "optional": {"by_dev_path", "by_uuid"},
    },
    "configure_storage_quota": {
        "required": {"device_name", "dev_path", "quota_limit_bytes"},
        "optional": {"quota_rotate"},
    },
    "storage_task_submit": {
        "required": {"device_name", "action", "dev_path"},
        "optional": {"sync", "files"},
    },
    "storage_task_status": {
        "required": {"device_name", "action", "dev_path"},
        "optional": set(),
    },
    "storage_task_cancel": {
        "required": {"device_name", "action", "dev_path"},
        "optional": set(),
    },
}

File: recamera-intellisense-mcp/src/api/storage.rs
use anyhow::{bail, Context, Result};
use serde_json::{json, Value};

use crate::api::expect_ok;
use crate::api_client::ApiClient;
use crate::types::{
    normalize_storage_action, DeviceRecord, RelayStatus, StorageSlot, StorageTaskHistory,
};

// MARK: Paths

const PATH_STATUS: &str = "/cgi-bin/entry.cgi/record/storage/status";
const PATH_CONFIG: &str = "/cgi-bin/entry.cgi/record/storage/config";
const PATH_CONTROL: &str = "/cgi-bin/entry.cgi/record/storage/control";

// MARK: Slot status + selection

pub async fn get_status(client: &ApiClient, device: &DeviceRecord) -> Result<Vec<StorageSlot>> {
    let data = client.get_json(device, PATH_STATUS, None).await?;
    let data_dir = data
        .get("sDataDirName")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    Ok(data
        .get("lSlots")
        .and_then(|v| v.as_array())
        .map(|arr| arr.iter().map(|s| parse_slot(s, &data_dir)).collect())
        .unwrap_or_default())
}

fn parse_slot(s: &Value, data_dir: &str) -> StorageSlot {
    let get_str = |k: &str| s.get(k).and_then(|v| v.as_str()).unwrap_or("").to_string();
    let get_bool = |k: &str| s.get(k).and_then(|v| v.as_bool()).unwrap_or(false);
    let get_i64 = |k: &str| s.get(k).and_then(|v| v.as_i64()).unwrap_or(0);
    let opt_str = |k: &str| s.get(k).and_then(|v| v.as_str()).map(String::from);
    StorageSlot {
        dev_path: get_str("sDevPath"),
        mount_path: get_str("sMountPath"),
        removable: get_bool("bRemovable"),
        internal: get_bool("bInternal"),
        label: opt_str("sLabel"),
        uuid: opt_str("sUUID"),
        fs_type: opt_str("sType"),
        selected: get_bool("bSelected"),
        enabled: get_bool("bEnabled"),
        syncing: get_bool("bSyncing"),
        writing: get_bool("bWriting"),
        rotating: get_bool("bRotating"),
        state_code: get_i64("eState"),
        state: get_str("sState"),
        size_bytes: get_i64("iStatsSizeBytes"),
        free_bytes: get_i64("iStatsFreeBytes"),
        quota_min_recommend_bytes: get_i64("iQuotaMinimumRecommendBytes"),
        quota_preserved_bytes: get_i64("iQuotaPreservedBytes"),
        quota_used_bytes: s.get("iQuotaUsedBytes").and_then(|v| v.as_i64()),
        quota_limit_bytes: get_i64("iQuotaLimitBytes"),
        quota_rotate: get_bool("bQuotaRotate"),
        data_dir: data_dir.to_string(),
    }
}

/// Select the slot to enable. Pass both selectors empty to disable all slots.
pub async fn set_selection(
    client: &ApiClient,
    device: &DeviceRecord,
    by_dev_path: &str,
    by_uuid: &str,
) -> Result<()> {
    let select = if by_dev_path.is_empty() && by_uuid.is_empty() {
        Value::Null
    } else {
        json!({ "sByDevPath": by_dev_path, "sByUUID": by_uuid })
    };
    let payload = json!({ "dSelectSlotToEnable": select });
    let resp = client
        .post_json(device, PATH_CONFIG, None, Some(&payload))
        .await?;
    expect_ok(&resp, "set storage selection")
}

// MARK: Sync control (CONFIG / RELAY / RELAY_STATUS / UNRELAY + sync variants)

pub async fn control_config(
    client: &ApiClient,
    device: &DeviceRecord,
    dev_path: &str,
    quota_limit_bytes: i64,
    quota_rotate: bool,
) -> Result<()> {
    let payload = json!({
        "sTaskType": "SYNC",
        "sAction": "CONFIG",
        "sSlotDevPath": dev_path,
        "dSlotConfig": {
            "iQuotaLimitBytes": quota_limit_bytes,
            "bQuotaRotate": quota_rotate,
        },
    });
    let resp = client
        .post_json(device, PATH_CONTROL, None, Some(&payload))
        .await?;
    expect_ok(&resp, "configure storage quota")
}

pub async fn control_relay_open(
    client: &ApiClient,
    device: &DeviceRecord,
    dev_path: &str,
) -> Result<RelayStatus> {
    let resp = relay_call(client, device, "RELAY", dev_path).await?;
    extract_relay_status(&resp)
}

pub async fn control_relay_status(
    client: &ApiClient,
    device: &DeviceRecord,
    dev_path: &str,
) -> Result<RelayStatus> {
    let resp = relay_call(client, device, "RELAY_STATUS", dev_path).await?;
    extract_relay_status(&resp)
}

pub async fn control_relay_close(
    client: &ApiClient,
    device: &DeviceRecord,
    dev_path: &str,
) -> Result<()> {
    let resp = relay_call(client, device, "UNRELAY", dev_path).await?;
    expect_ok(&resp, "close relay")
}

async fn relay_call(
    client: &ApiClient,
    device: &DeviceRecord,
    action: &str,
    dev_path: &str,
) -> Result<Value> {
    let payload = json!({
        "sTaskType": "SYNC",
        "sAction": action,
        "sSlotDevPath": dev_path,
    });
    let resp = client
        .post_json(device, PATH_CONTROL, None, Some(&payload))
        .await?;
    expect_ok(&resp, &format!("storage control {action}"))?;
    Ok(resp)
}

fn extract_relay_status(resp: &Value) -> Result<RelayStatus> {
    let r = resp
        .get("dRelayStatus")
        .context("missing dRelayStatus in response")?;
    Ok(RelayStatus {
        uuid: r
            .get("sRelayDirectory")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
        timeout: r.get("iRelayTimeout").and_then(|v| v.as_i64()).unwrap_or(0),
        timeout_remain: r
            .get("iRelayTimeoutRemain")
            .and_then(|v| v.as_i64())
            .unwrap_or(0),
    })
}

// MARK: Long-running actions (FORMAT / FREE_UP / EJECT / REMOVE)

pub async fn control_sync(
    client: &ApiClient,
    device: &DeviceRecord,
    action: &str,
    dev_path: &str,
    files: &[String],
) -> Result<Value> {
    let action = normalize_storage_action(action)
        .with_context(|| format!("unknown storage action '{action}'"))?;
    let mut payload = json!({
        "sTaskType": "SYNC",
        "sAction": action,
        "sSlotDevPath": dev_path,
    });
    if action == "REMOVE_FILES_OR_DIRECTORIES" {
        if files.is_empty() {
            bail!("REMOVE_FILES_OR_DIRECTORIES requires a non-empty 'files' list");
        }
        payload["lFilesOrDirectoriesToRemove"] = json!(files);
    }
    let resp = client
        .post_json(device, PATH_CONTROL, None, Some(&payload))
        .await?;
    expect_ok(&resp, &format!("storage {action} (sync)"))?;
    Ok(resp)
}

pub async fn control_async_submit(
    client: &ApiClient,
    device: &DeviceRecord,
    action: &str,
    dev_path: &str,
    files: &[String],
) -> Result<StorageTaskHistory> {
    let action = normalize_storage_action(action)
        .with_context(|| format!("unknown storage action '{action}'"))?;
    let mut payload = json!({
        "sTaskType": "ASYNC_SUBMIT",
        "sAction": action,
        "sSlotDevPath": dev_path,
    });
    if action == "REMOVE_FILES_OR_DIRECTORIES" {
        if files.is_empty() {
            bail!("REMOVE_FILES_OR_DIRECTORIES requires a non-empty 'files' list");
        }
        payload["lFilesOrDirectoriesToRemove"] = json!(files);
    }
    let resp = client
        .post_json(device, PATH_CONTROL, None, Some(&payload))
        .await?;
    expect_ok(&resp, &format!("submit async {action}"))?;
    Ok(resp)
}

pub async fn control_async_status(
    client: &ApiClient,
    device: &DeviceRecord,
    action: &str,
    dev_path: &str,
) -> Result<StorageTaskHistory> {
    let action = normalize_storage_action(action)
        .with_context(|| format!("unknown storage action '{action}'"))?;
    let payload = json!({
        "sTaskType": "ASYNC_STATUS",
        "sAction": action,
        "sSlotDevPath": dev_path,
    });
    let resp = client
        .post_json(device, PATH_CONTROL, None, Some(&payload))
        .await?;
    expect_ok(&resp, &format!("query async {action} status"))?;
    Ok(resp)
}

pub async fn control_async_cancel(
    client: &ApiClient,
    device: &DeviceRecord,
    action: &str,
    dev_path: &str,
) -> Result<StorageTaskHistory> {
    let action = normalize_storage_action(action)
        .with_context(|| format!("unknown storage action '{action}'"))?;
    let payload = json!({
        "sTaskType": "ASYNC_CANCEL",
        "sAction": action,
        "sSlotDevPath": dev_path,
    });
    let resp = client
        .post_json(device, PATH_CONTROL, None, Some(&payload))
        .await?;
    expect_ok(&resp, &format!("cancel async {action}"))?;
    Ok(resp)
}
